base/commands/serializer/java.go

```diff
@@ -0,0 +1,275 @@
+	if strings.HasSuffix(fieldType, "[]") {
+		return fieldType[0 : len(fieldType)-2]
+	}
+	return fieldType
+}
+
+func methodName(field TypeInfo, compactType string) string {
+	var mn string
+	if field.IsCustomClass {
+		mn = "compact"
+	} else {
+		mn = compactType
+	}
+	// capitalize fist letter
+	mn = strings.ToUpper(string(mn[0])) + mn[1:]
+	if field.IsArr {
+		return fmt.Sprintf("ArrayOf%s", arrayOf(mn))
+	}
+	return mn
+}
+
+func toJavaType(fieldType string) TypeInfo {
+	var ti TypeInfo
+	var ok bool
+	if strings.HasSuffix(fieldType, "[]") {
+		base := arrayOf(fieldType)
+		ti.IsArr = true
+		ti.Type, ok = javaTypes[base]
+		if ok {
+			ti.FullType = ti.Type + "[]"
+		} else {
+			ti.IsCustomClass = true
+			ti.Type = base
+			ti.FullType = fieldType
+		}
+	} else {
+		ti.Type, ok = javaTypes[fieldType]
+		if !ok {
+			ti.Type = fieldType
+			ti.IsCustomClass = true
+		}
+		ti.FullType = ti.Type
+	}
+	return ti
+}
+
+func fieldTypeAndNames(cls Class) string {
+	var sb strings.Builder
+	for _, f := range cls.Fields {
+		ti := toJavaType(f.Type)
+		sb.WriteString(fmt.Sprintf("%s %s, ", ti.FullType, f.Name))
+	}
+	s := sb.String()
+	if len(s) >= 2 {
+		s = s[:len(s)-2]
+	}
+	return s
+}
+
+func fieldNames(class Class) string {
+	var sb strings.Builder
+	for _, f := range class.Fields {
+		sb.WriteString(fmt.Sprintf("%s, ", f.Name))
+	}
+	content := sb.String()
+	return content[:len(content)-2]
+}
+
+func hashcodeBody(cls Class) string {
+	const indentation = "        "
+	var content string
+	var isTempDeclared bool
+	for _, field := range cls.Fields {
+		var line string
+		fn := field.Name
+		switch field.Type {
+		case "boolean":
+			line = fmt.Sprintf("result = 31 * result + (%s ? 1 : 0);", fn)
+		case "int64":
+			line = fmt.Sprintf("result = 31 * result + (int) (%s ^ (%s >>> 32));", fn, fn)
+		case "float32":
+			line = fmt.Sprintf("result = 31 * result + (%s != +0.0f ? Float.floatToIntBits(%s) : 0);", fn, fn)
+		case "float64":
+			if !isTempDeclared {
+				line = "long temp;\n"
+				isTempDeclared = true
+			}
+			line += fmt.Sprintf(`%stemp = Double.doubleToLongBits(%s);
+%sresult = 31 * result + (int) (temp ^ (temp >>> 32));`, indentation, fn, indentation)
+		default:
+			if strings.HasSuffix(field.Type, "[]") {
+				line = fmt.Sprintf("result = 31 * result + Arrays.hashCode(%s);", field.Name)
+			} else if _, ok := fixedSizeTypes[field.Type]; ok {
+				line = fmt.Sprintf("result = 31 * result + (int) %s;", field.Name)
+			} else {
+				line = fmt.Sprintf("result = 31 * result + Objects.hashCode(%s);", field.Name)
+			}
+		}
+
+		content += "        " + line + "\n"
+	}
+	return content
+}
+
+func toStringBody(field Field) string {
+	if strings.HasSuffix(field.Type, "[]") {
+		return fmt.Sprintf("                + \", %s=\" + Arrays.toString(%s)\n", field.Name, field.Name)
+	}
+	return fmt.Sprintf("                + \", %s=\" + %s\n", field.Name, field.Name)
+}
+
+func toStringBodyFirst(field Field) string {
```

Comment:
I am not sure as well, maybe it is better to ask its author.

Reply:
`toStringBody` and `toStringBodyFirst` seem to do almost the same thing, except `toStringBody` has `, ` at the beginning of the lines.
Would it make sense to add an `index` parameter to `toStringBody`?
If `index == 0` the prefix would be blank otherwise it would be `, `.
That would make the following template:
```
{{range $index, $field := .Cls.Fields}}{{ if eq $index 0}}{{ toStringBodyFirst $field }}{{else}}{{ toStringBody $field }}{{end}}{{end}}
```
to:
```
{{range $index, $field := .Cls.Fields}}{{ toStringBodyFirst $index $field }}{{end}}
```

In that case we can also remove the `toStringBodyFirst` function.

Also, how about renaming `toStringBody` to `generateStringMethodBody`?